pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	r.Service = services.NewBasicService(r.start, r.run, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffset(ctx)
+	if err != nil {
+		return err
+	}
+	level.Info(r.logger).Log("msg", "resuming consumption from offset", "offset", offset)
+
+	r.client, err = r.newKafkaReader(offset, r.reg)
+	if err != nil {
+		return errors.Wrap(err, "creating kafka reader client")
+	}
+	return nil
+}
+
+func (r *PartitionReader) run(ctx context.Context) error {
+	consumeCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
+		}
+		level.Debug(r.logger).Log("msg", "fetched records", "num_records", fetches.NumRecords())
+
+		r.recordFetchesLag(fetches)
+		r.consumeFetches(consumeCtx, fetches)
+		r.commitFetches(consumeCtx, fetches)
+	}
+
+	return nil
+}
+
+func collectFetchErrs(fetches kgo.Fetches) (_ error) {
+	mErr := multierror.New()
+	fetches.EachError(func(s string, i int32, err error) {
+		// TODO handle errors properly, there can be some error we can ignore and some errors for which we have to reset the kafka client. See docs on EachError
+		mErr.Add(err)
+	})
+	return mErr.Err()
+}
+
+func (r *PartitionReader) commitFetches(ctx context.Context, fetches kgo.Fetches) {
+	committed, err := kadm.NewClient(r.client).CommitOffsets(ctx, consumerGroup, kadm.OffsetsFromFetches(fetches))
+	if err != nil {
+		level.Error(r.logger).Log("msg", "encountered error while committing offsets", err)
+	} else {
+		committedOffset, _ := committed.Lookup(r.kafkaTopic, r.partition)
+		level.Debug(r.logger).Log("msg", "committed offset", "offset", committedOffset.Offset.At)
+	}
+}
+
+func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetches) {
+	fetches.EachRecord(func(record *kgo.Record) {
+		level.Debug(r.logger).Log("msg", "fetched record", "offset", record.Offset)
+		defer prometheus.NewTimer(r.metrics.processingTime).ObserveDuration()
+
+		err := r.consumer.Consume(ctx, mapRecord(record))
+		if err != nil {
+			level.Error(r.logger).Log("msg", "encountered error processing record; skipping", "offset", record.Offset, "err", err)
+			// TODO abort ingesting & back off if it's a server error, ignore error if it's a client error
+		}
+	})
+}
+
+func mapRecord(record *kgo.Record) Record {
+	return Record{
+		Content:  record.Value,
+		TenantID: string(record.Key),
+	}
+}
+
+func (r *PartitionReader) recordFetchesLag(fetches kgo.Fetches) {
+	processingStart := time.Now()
+	fetches.EachRecord(func(record *kgo.Record) {
+		r.metrics.receiveDelay.Observe(processingStart.Sub(record.Timestamp).Seconds())
```

Comment:
We need to double check if `Timestamp` is correctly supported by the backend we use. I vaguely remember there were some issues with it.

Reply:
i verified that timestamp is correctly propagated on the backends we use